src/main/java/ch/njol/skript/sections/EffSecShoot.java

```diff
@@ -0,0 +1,298 @@
+					} else {
+						finalProjectile = entityData.spawn((Location) shooter);
+					}
+				}
+				if (finalProjectile != null) {
+					finalProjectile.setVelocity(vector);
+					lastSpawned = finalProjectile;
+				}
+			}
+		}
+
+		return super.walk(event, false);
+	}
+
+	private static <E extends Entity> void set(Entity entity, EntityData<E> entityData) {
+		//noinspection unchecked
+		entityData.set((E) entity);
+	}
+
+	private CaseUsage getCaseUsage(Boolean isProjectile, Boolean useWorldSpawn, Boolean hasTrigger) {
+		if (!isProjectile) {
+			if (!hasTrigger)
+				return CaseUsage.NOT_PROJECTILE_NO_TRIGGER;
+			return CaseUsage.NOT_PROJECTILE_TRIGGER;
+		}
+		if (!useWorldSpawn) {
+			if (!hasTrigger)
+				return CaseUsage.PROJECTILE_NO_WORLD_NO_TRIGGER;
+			if (launchWithBukkitConsumer != null)
+				return CaseUsage.PROJECTILE_NO_WORLD_TRIGGER_BUKKIT;
+			return CaseUsage.PROJECTILE_NO_WORLD_TRIGGER;
+		}
+		if (!hasTrigger)
+			return CaseUsage.PROJECTILE_WORLD_NO_TRIGGER;
+		return CaseUsage.PROJECTILE_WORLD_TRIGGER;
+	}
+
+	private Consumer<? extends Entity> afterSpawn(Event event, EntityData<?> entityData, @Nullable LivingEntity shooter) {
+		return entity -> {
+			if (entity instanceof Fireball fireball)
+				fireball.setShooter(shooter);
+			else if (entity instanceof Projectile projectile && shooter != null) {
+				projectile.setShooter(shooter);
+				set(projectile, entityData);
+			}
+			ShootEvent shootEvent = new ShootEvent(entity, shooter);
+			lastSpawned = entity;
+			Variables.setLocalVariables(shootEvent, Variables.copyLocalVariables(event));
+			TriggerItem.walk(trigger, shootEvent);
+			Variables.setLocalVariables(event, Variables.copyLocalVariables(shootEvent));
+			Variables.removeLocals(shootEvent);
+		};
+	}
+
+	/**
+	 * MC 1.19 uses Bukkit Consumer for LivingEntity#launchProjectile instead of Java Consumer
+	 */
+	@SuppressWarnings("deprecation")
+	private org.bukkit.util.Consumer<? extends Entity> afterSpawnBukkit(Event event, EntityData<?> entityData, @Nullable LivingEntity shooter) {
+		return entity -> {
+			if (entity instanceof Fireball fireball)
+				fireball.setShooter(shooter);
+			else if (entity instanceof Projectile projectile && shooter != null) {
+				projectile.setShooter(shooter);
+				set(projectile, entityData);
+			}
+			ShootEvent shootEvent = new ShootEvent(entity, shooter);
+			lastSpawned = entity;
+			Variables.setLocalVariables(shootEvent, Variables.copyLocalVariables(event));
+			TriggerItem.walk(trigger, shootEvent);
+			Variables.setLocalVariables(event, Variables.copyLocalVariables(shootEvent));
+			Variables.removeLocals(shootEvent);
+		};
+	}
+
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		return "shoot " + types.toString(event, debug) + " from " + shooters.toString(event, debug) +
+			(velocity != null ? " at speed " + velocity.toString(event, debug) : "") +
+			(direction != null ? " " + direction.toString(event, debug) : "");
```

Comment:
When you're appending strings over multiple lines, the pluses should go at the start of the new line.
```suggestion
		return "shoot " + types.toString(event, debug) + " from " shooters.toString(event, debug) +
			+ (velocity != null ? " at speed " + velocity.toString(event, debug) : "")
			+ (direction != null ? " " + direction.toString(event, debug) : "");
```
This convention comes from mathematics because it makes it easier to determine the relation between elements in a displayed formula. If you're interested to learn about it, Donald Knuth explains it in _The TeXBook_ (pp. 195-196).